In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [3]:
df_prods = pd.read_csv(r'/Users/matthewerrington/08-2020 Instacart Basket Analysis/02 Data/Original Data/products.csv', index_col = False)

In [4]:
df_ords = pd.read_csv(r'/Users/matthewerrington/08-2020 Instacart Basket Analysis/02 Data/Prepared Data/orders_wrangled.csv', index_col = False)

In [6]:
df_ords.describe()

Unnamed: 0      order_id       user_id  user_order_count  \
count  3.421083e+06  3.421083e+06  3.421083e+06      3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05      1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04      1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00      1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04      5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05      1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05      2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05      1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [7]:
# Create a dataframe

df_test = pd.DataFrame()

In [8]:
# Create a mixed type column

df_test['mix'] = ['a', 'b', 1, True]

In [9]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [11]:
# Check for mixed types

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [12]:
# Fix mised type

df_test['mix'] = df_test['mix'].astype('str')

# Missing Values

In [13]:
# Finding missing values

In [14]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [15]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [16]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

# Addressing Missing Values

In [17]:
# Create a new variable that acts like a flag based on the missing value
# Impute the value with the mean or median of the column
# Remove or filter out the missing data

In [19]:
## Replace missing value with mean/average
# df['column with missings'].fillna(mean value, inplace=True)


In [20]:
## Replace missing value with median
# df['column with missings'].fillna(median value, inplace=True)

In [22]:
df_prods.shape

(49693, 5)

In [24]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [25]:
df_prods_clean.shape

(49677, 5)

# Duplicates

In [29]:
# Finding duplicates

In [27]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [28]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [30]:
# Addressing duplicates

In [32]:
df_prods_clean.shape

(49677, 5)

In [35]:
# drop_duplicates removes duplicates from the dataframe
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [36]:
df_prods_clean_no_dups.shape

(49672, 5)

In [38]:
## Export Checked Dataframe
## Remove comment when ready to export!
# df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'prods_checked.csv'))

# Exercise 4.5

In [46]:
# Question 2
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

It's interesting to note that the max price is $99,999
Unlikely figure for Instacart purchase unless it's vintage wine...

In [41]:
# Question 3
for col in df_ords.columns.tolist():
  mixed_type = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[mixed_type]) > 0:
    print (col)

No mixed-type data revealed for df_ords dataframe

In [44]:
# Question 5
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
user_order_count               0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

It makes since that days_since_prior_order would have missing values.
Not every customer will have multiple orders.

In [47]:
# Question 6

I will not replace or remove the missing values for days_since_prior_order. The missing value shows which customers may need marketing, etc. 

In [49]:
# Question 7
df_ords_dups = df_ords[df_ords.duplicated()]

In [50]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, user_order_count, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

No duplicates found in the orders dataframe.

In [51]:
# Question 8

No duplicates found to address.

In [54]:
# Question 9
path = r'/Users/matthewerrington/08-2020 Instacart Basket Analysis'
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'prods_checked.csv'))

In [55]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'ords_checked.csv'))